### PR TITLE
Implement IPAddressFilterSet search

### PR DESCRIPTION
**Pull request by andrew-wm-arthur**
fix for #287

### CODE REVIEW

File: nautobot/ipam/querysets.py [hunk 1]
```diff
@@ -184,6 +183,43 @@ def get_queryset(self):
...
+        return self.filter(
+            Q(dns_name__icontains=search)
+            | Q(description__icontains=search)
+            | Q(host__lte=broadcast, host__gte=network.network)  # same as `net_host_contained()`
+        )
+
+    def _parse_as_network_string(self, search):
+        """
+        Attempts to parse a (potentially incomplete) IPAddress and return an IPNetwork.
+        eg: '10.10' should be interpreted as netaddr.IPNetwork('10.10.0.0/16')
```
Comment (glennmatthews) [inline nautobot/ipam/querysets.py:204]: Don't we need similar logic for IPv6 addresses? Will almost certainly be more complicated due to the myriad of ways to enter a full or partial v6 address -  "2001:db8:", "2001:0db8::", "2001:DB8::1", etc.

**Review by glennmatthews — COMMENTED** [review]
Thanks for taking this on!

File: nautobot/ipam/querysets.py [hunk 2]
```diff
@@ -184,6 +190,97 @@ def get_queryset(self):
...
+            return netaddr.IPNetwork(ver_map[version])
+
+    def _parse_ipv6(self, value):
+        """IPv6 addresses are 8, 16-bit fields."""
+
+        # Get non-empty octets from search string
+        hextets = value.split(":")
+
+        # Before we normalize, check that final value is a digit.
+        if hextets[-1].alnum():
```
Comment (jathanism) [inline nautobot/ipam/querysets.py:254]: Whoops this should be `isalphanum()`
```suggestion
        if hextets[-1].alphanum():
```

**Review by jathanism — APPROVED** [review]
Not bad if I do say so myself, which I did. ;P

File: nautobot/ipam/querysets.py [hunk 3]
```diff
@@ -184,6 +190,97 @@ def get_queryset(self):
...
+            return call_map[version](search)
+
+        except netaddr.core.AddrFormatError:
+            ver_map = {4: "0/32", 6: "::/128"}
+            return netaddr.IPNetwork(ver_map[version])
+
+    def _parse_ipv6(self, value):
+        """IPv6 addresses are 8, 16-bit fields."""
+
+        # Get non-empty octets from search string
```
Comment (jathanism) [inline nautobot/ipam/querysets.py:250]:
```suggestion
        # Get non-empty hextets from search string
```

File: nautobot/ipam/querysets.py [hunk 4]
```diff
@@ -184,6 +190,97 @@ def get_queryset(self):
...
+        hextets = value.split(":")
+
+        # Before we normalize, check that final value is a digit.
```
Comment (jathanism) [inline nautobot/ipam/querysets.py:253]:
```suggestion
        # Before we normalize, check that final value is alphanumeric.
```

File: nautobot/ipam/tests/test_filters.py [hunk 5]
```diff
@@ -754,6 +754,39 @@ def setUpTestData(cls):
...
+            "2001:": 5,
+            "2001::": 5,
+            "2001:db8:": 5,
+            "2001:db8::": 5,
+            "2001:db8::/64": 5,
+            "2001:db8::2": 1,
+            "2001:db8:0:2": 0,
+            "fe80": 0,
+            "fe80::": 0,
+            "foo.bar": 0,
```
Comment (jathanism) [inline nautobot/ipam/tests/test_filters.py:783]:
```suggestion
            "ipaddress-a": 2,
            "foo": 0,
            "foo.bar": 0,
            # network searches
            "": 10,
            "10": 5,
            "10.": 5,
            "10.0": 5,
            "10.0.0.4": 1,
            "10.0.0.4/24": 1,
            "11": 0,
            "11.": 0,
            "11.0": 0,
            "10.10.10.0/24": 0,
            "2001": 5,
            "2001:": 5,
            "2001::": 5,
            "2001:db8:": 5,
            "2001:db8::": 5,
            "2001:db8::/64": 5,
            "2001:db8::2": 1,
            "2001:db8:0:2": 0,
            "fe80": 0,
            "fe80::": 0,
```

**Review by lampwins — APPROVED** [review]

**Review by glennmatthews — APPROVED** [review]
No doubt there are probably IPv6 corner cases we're missing, but we can squash those individually as and when they come up. Nice work!

**Pull request merged**